### ISSUE TITLE
Flight - Improve GPS Position Hold (GPS PH)

### ISSUE THREAD

**Issue by Akfreak** [cleanflight/cleanflight#197]
This folder is created per Hydra's request, "<hydra> perhaps add some of your ideas to a new github issue tracker issue
<hydra> if you create one called ‘gps improvements’
<hydra> and then add comments to it for each idea, then we can discount or progress each idea you have.

First we need to use the best GPS, Compass, and IMU (combined as a single GPS tracking and data collection unit.) which the Naze 32 (and like boards) can support.  Using GPS coordinates alone for hold isn't gong to cut it.  We need to consider the sensor data from the GPS, compass (gps compass) and IMU to write code that can use all of the data together to control drift. By fusing data from all three devices, we can definitely know if the craft is drifting from the GPS coordinates obtained when you enter GPS mode or GPS hold mode. I hope I am making sense. 

Once we are able to lock the crafts positions, we can do all sorts of neat stuff. For example, we can record the gps data, IMU data, and compass data, (ground track, plus known past IMU and compass data that has been collected during flight) along with the home position data, and we can return to home with no active GPS. We could write a routine that takes all the data and creates a path back to home (shortest distance, raising up in elevation to clear obstacles) and fly it based on the stored sensor data. Like a backup RTH if your GPS ever failed. Add Lidar for collision avoidance you might have a cool autonomous rig 

The nose in or nose out on RTH is not that difficult when you have the correct position data and a machine that can hold on GPS coordinates.. In the GUI you could just have a check box face in or away on RTH. Then have it fly home normally and before it descend have it rotate to specified orientation and land ( just subtract or add 180° to known data collected on takeoff to the yaw axis ) 

Also in GPS hold mode, we could use the throttle stick for height hold. Yep we map the throttle throws, and have the flight controller use GPS (full package GPS, Compass and IMU data) and baro data in conjunction with the throttle. When in GPS flight mode, the throttle stick just becomes the main control for the GPS hold feature. To auto take off, just put the throttle stick at 50% (+ or - 3-5 %) and the craft takes off and holds 3 to 5 meters off the ground (we could control throttle expo easily). Then if you go over the 50% threshold the machine climbs, if you go below 50% throttle, the machine descends.  This would be perfect for AP. If you want to climb up a little, give a little throttle above 50% then u go up. When you return to 50% throttle and it locks in GPS and Altitude hold . You could still use Yaw, Pitch and Roll (with set limits, 25° to 30° max) to move about, but you would not climb or descend. To go down, just reduce throttle below 50% and go down. When the throttle stick is at center the machine resumes GPS lock. It would be a cool way to fly. 

Before we can do all this fun stuff, we need solid lock. IMO, this will only happen by adding a standalone GPS with additional compass, and IMU to collect sensor data. It would also be nice to add a micro SD card for Black Box flight data that monitors all systems, but this is a whole different issue.  

I hope my ideas are what you are looking for. If not please can you give examples of what you are asking for. Thanks   Tom

**Comment by hydra** [comment]
Getting a solid PH is key here.  Let's keep this ticket focused on improving GPS PH.  Renaming to reflect since this ticket is too vague.

**Comment by matbogdan79** [comment]
I am not an expert in GPS and GPS PH, but read that this is already done in other projects with great success. Isn't easier to get inspiration from the code already made and tested ?

**Comment by hydra** [comment]
Yes, we can leverage existing code.

**Comment by Redshifft** [comment]
Good GPS PH is not going to be down to only GPS accuracy, I think that's a small part.I know what a Naza can do with 7 sats ;)
I think CF also needs to automatically switch all required sensors when a GPS function is selected also is there some sort of very high 3d lock sat count threshold in place ? I'm getting inconsistant reslts in tests here :(

**Comment by sophana** [comment]
Hi. 

I'm also no expert in gps, but have some idea about how it could work. 

The gps coordinates frequency are quite low (5 to 18 per second). They can have errors. I assume that when you average several positions (low pass temporal filter), the error decreases.

The accelerometers give high frequency updates. Position is an integral of acceleration. But the integral of acceleration drifts: the low frequency of this estimated position is wrong, but the high frequency is much better.

So the idea is to combine both to get a good position estimation at any moment.

To be able to integrate acceleration you need a good "heading" estimation from both gyros and magnetometers, with the same high/low frequency separation.

Once you have this, you can integrate accelerometer associated with heading information. This integration tends to drift, so you use a high pass filter on this.
The gps coordonates are first going though a low pass filter, and then interpolated. You get a low frequency interpolated position. I'm assuming that the low-pass filter will decrease the position error.

You then combine this low frequency position from gps with the high frequency position from the accelerometers, and I think you get a quite good position estimation which you can use to hold.

The cut frequency of both the high pass and low pass filters should be matched.

This high/low frequency separation should also apply with a very similar technique on the gyro/magnetometer info.

Feel free to comment.

**Comment by digitalentity** [comment]
@hydra does anybody working on this already? I may be able to improve GPS PH using accelerometer-based speed calculations in my free time.

**Comment by dkisselev** [comment]
@digitalentity, I've been poking around and probably won't have time to write code, but if you happen to find the time (I know how hard it is), here are a few useful links I've found:

The "proper" way to do dual-sensor stabilization is through a Kalman filter. It's similar to a PID in that it adds error values up to figure out a more "true" position. 

Here's a really good video outlining what it looks like when it works well: https://www.youtube.com/watch?v=nNWWLJZRxAU

In the description for the video he posted a few good links about the math behind the filter. The second link is really really good.

If I work on the rewrite, I'd probably start by cleaning up all the MultiWii legacy stuff. Consolidate the functions down into some good semantic units, clean up variables, etc. Get the core GPS stuff working, then add in the error correction.

**Comment by digitalentity** [comment]
@dkisselev I was thinking about re-writing navigation code from scratch, re-using only a few lines, cleaning things up, adding abiity to get error correction every looptime, not every GPS update. After that the next step would be adding sensor fusion to get more frequent and precise updates.

Kalman filter is good, but unfortunately I don't have the knowledge to properly implement Kalman-based GPS navigation.

I was thinking about integrating copter acceleration (raw acceleration - gravity) twice to get relative movement, convert it to LAT/LON movement and use a complementary filter to fuse this with GPS coordinates. Integratged acceleration will provide fast response, GPS will compensate for drift.

What do you think?

**Comment by dkisselev** [comment]
Yeah, I was thinking the same thing (about re-writing), but wasn't sure how deep you wanted to go into the project.

The second link in the video I linked has some really good code for implementing a 2D kalman in Python. I agree that the math seems tough, but looking at the code it seems simple enough to implement. My biggest worry with it would be figuring out how many CPU cycles the filter will burn, it requires you to invert a few matrices per cycle, which is a pretty expensive operation.

Sounds like your method could work pretty well. I think the ultimate solution is to work the GPS rewrite in a way that allows passing the "final" co-ordinates through a definable filter function, where we can implement all these things that we might want.

**Comment by digitalentity** [comment]
I'll try to keep the filter as modular as possible, so we could choose what filter we'll be using during compilation. I would probably stick with complementary as it is simple and inexpensive in terms of cpu cycles. Kalman might just use too many cycles on STM32F1. F3-based platforms (Sparky) could use Kalman if somebody (might be myself) will implement it.
Feel free to watch the progress in my fork here: https://github.com/digitalentity/cleanflight/tree/gps-rewrite but please don't expect instant results :blush:

**Comment by dkisselev** [comment]
I don't have a GPS on my quad yet (I got sent the wrong naze baord and ended up without a baro/compass, and don't want to pay for a new one or jerry-rig the sensors on), but I'll contribute what I can :)

**Comment by digitalentity** [comment]
I am having some trouble integrating accelerometer to velocity, probably due to some problems in imu.c, will look into this later. So far I am reworking and cleaning up gps code from legacy stuff.

**Comment by ledvinap** [comment]
@digitalentity : You should probably take blackbox logs, load them into Octave (or Matlab is you have it) and try to get some useful data out. It would be much easier to implement algorithm that is tested on blackbox data ..

**Comment by digitalentity** [comment]
@ledvinap, thanks for the great idea! This way I'll have the ability test and debug the new algorythms without having to actually fly a copter

**Comment by ledvinap** [comment]
@digitalentity : this may save you some time if you choose Octave. It parses blackbox header and saved data into variables (prefixed vith `d_`). Must be sourced, use `source <file.m>` or `octave --persist <file.m>`

```
%%% Emacs, this is -*- octave -*-

pkg load signal;

file = fopen("LOG00092.01.csv");
%file = fopen("tracernz/LOG00070.01.csv");

hdrs = strtrim(strsplit(fgetl(file), ","));

data = dlmread(file, ",");
data_def = struct ();
for i = 1 : size(hdrs,2)
  def = regexp(hdrs(i), '^(?<var>\w+)(\[(?<idx>\d+)\])?(\s*\([\w/]+\))?\s*$', "names"){1,1};
  if (isempty(def.idx))
    idx = 1;
  else
    idx = str2num(def.idx) + 1;
  endif
  if(!isfield(data_def, def.var))
    data_def.(def.var) = [];
  endif
  data_def.(def.var)(idx) = i;
endfor

for [val, key] = data_def
  eval(strcat('d_',key,' = data(:,', mat2str(val),');'));
endfor

% d_time = d_time - d_time(1);  % shift time to start from zero
Ts = 0.004;                                 % put looptime here
Fs = 1/Ts;
Ws = 2*pi*Fs;

t=d_time/1000000;

% plot gyro axis X and 40Hz 2nd order butterworth filtered version of it
[b,a] = butter(2, 2*40/Fs);
% use filtfilt(b,a,data) to compensate filter delay. 
plot(t, d_gyroData(:, 1), t, filter(b, a, d_gyroData(:, 1));

% plot spectrogram of gyro X data
step=ceil(100*Fs/1000);
window=ceil(200*Fs/1000);
fftn = 2^nextpow2(window);
specgram(d_gyroData(:,1), fftn, Fs, window, window-step);
```

**Comment by Redshifft** [comment]
Sounding good for the future :)
LOL, I actually got the existing code to RTL today with a camera to prove it !
Basically it's mostly useless as there is no automatic height control, also operation is flaky, which I put down to it needing more Sats to function than I regularly pick up.
I can't code but I'm happy to test any early development builds ;)

**Comment by dkisselev** [comment]
@Redshifft I believe height control is expected from the baro component. So you have to set your modes up to switch to baro mode and RTH under the same aux input.

Agreed that that's sketchy though.

**Comment by hydra** [comment]
@dkisselev baro is improved in 1.9.0

**Comment by dkisselev** [comment]
@hydra I meant more that the GPS component does not maintain height. So unless you explicity turn baro hold on, your quad could RTH into the ground if the angle that it goes at with the user's throttle setting causes it to lose altitude.

I don't have a GPS though, so this is just speculative based on what I see in the code.

**Comment by digitalentity** [comment]
@dkisselev @hydra from what I see in the code, we should probably drop separate navigation & altitudehold modules and re-create a complete position estimation and navigation. This position estimation should include altitude fused together from all available sources (baro, sonar, gps). I am currently thinking of porting INS/GPS code from Taulabs (https://github.com/TauLabs/TauLabs/blob/next/flight/Libraries/insgps13state.c) over to CF, however I am not sure if F1 targets would be able to handle the math, as calculations are comlex.

**Comment by dkisselev** [comment]
I actually think that the current "navigation" component does too much. 

I feel like all the GPS-related stuff should be pulled out into a GPS module and navigation should only be responsible for handling the current flight mode (making sure to enable baro hold when flying to a waypoint...), the waypoint stack (whcih will only hold a single RTH waypoint for now), and any of those related commands.

The navigator would then call on GPS to get the current position, do the Nav PID calculations, and tell the flight PID's how much to tilt in which direction. Then it'll call on baro to get the current height, do that calculation, run the baro PID to keep the thing from hitting the ground, and tell the flight PID to bump the throttle.

That way it's easy to add that accellerometer-based navigation stuff.

It looks like taulabs actually implement a really complicated Kalman filter to do their thing. It looks like they're trying to do it in three dimensions right away, which might be why it's so complicated. 

I think at this point our baro/sonar hold is good enough (~1m variance) that we can keep that separate, which simplifies our implementation of the filter.

**Comment by digitalentity** [comment]
I like this idea. Each sensor should use it's own noise reduction and smoothing (as it is already done in barometer sensor).

**Comment by dkisselev** [comment]
Yep, then nav is only responsible for taking lat/long, height, and direction values as determined by other sensors (any sensors), amalgamating multiple sensors into a single "master position in space" parameter, finding the lat/long difference between the current master position and waypoint, and setting the appropriate attack angle to fly there.

**Comment by tracernz** [comment]
I don't think the Taulabs code is really portable to CF without major changes to the whole codebase and you will need the equivalent math knowledge of an engineering degree. They have a proper matrix based attitude and position estimation model which fuses all of the sensors (gyro, accel, baro, mag, gps, etc.) and uses the combination to determine the crafts position and orientation in 3d space and using complementary sensors (e.g. gyro & accel) to help filter noise and bias out of each other.

**Comment by Redshifft** [comment]
@dkisselev   Not the normal baro height control, an automatic climb say +5 meters then RTL is missing. Infact there is actually no automatic height control other than your throttle stick.

**Comment by Redshifft** [comment]
@hydra Tests were on ver 1.7.1  I did actually wonder if the baro on the board was even supported in Cleanflight ? It's an AfroNaze Mini, I could not see the sensor in the code (only the earlier version) BMP805, I have the BMP180 fitted

**Comment by digitalentity** [comment]
I've implemented Kalman-based GPS filtering but don't have any data to test it on. The weather in my city is bad for flying so I ask for help. If anybody can give me some blackbox logs with gps data in them, I'll test new GPS filtering algorythm. Probably it would be enough to make poshold working better even with current navigation code.

**Comment by csurf** [comment]
What's the status&progress on the GPS/RTH features? Is "RTH minimum height" being worked on (PH and climb until a pre-configured minimum height is reached & then initiate RTH) ?

Would it be possible to submit a temporary patch in order to get this to work in the meantime (at least until a proper solution is developed)? I'm thinking something that's based on the 2.4 MW code where the MR is placed into GPSHOLD and begins a slow, progressive climb until RTH height is reached, at which point it switches over to RTH. thoughts?

**Comment by digitalentity** [comment]
@csurf, I am working on completely re-writing navigation system. New system will replace current navigation and althold with new navigation system with all bells and whistles like RTH minimum height, GPS assisted flight, waypoint navigation etc. However the new system is far from complete. IMHO any temporary solution is welcome, we all would benefit from it.

**Comment by marksev1** [comment]
Digitalentity after you finalize the WP stuff would be cool if you would integrate PatrikEs fixedwing-nav stuff thus allowing also WP for fixed wing on stm32f103 :)

**Comment by digitalentity** [comment]
New nav code replaces current gps navigation and althold. Changes are vast and touch many parts of CF, that's why I am focused on getting this working on multirotors first. But after this gets implemented and tested I'll look into fixed-wing navigation: RTH, PH (loiter), WP etc

**Comment by marksev1** [comment]
Great, thanks.

**Comment by OshCFII** [comment]
I'm not sure of the likely time required to complete a GPS code re-write, but I'd guess it is measured in months and not weeks.  Would it make sense to try to integrate some of the GPS Position and Hold capabilities from CrashPilot1000's Harakari code?  From what I've read, it has the best PH and RTH of any firmware that runs on the Naze32 type boards.

I know that CrashPilot1000 has dropped by and offered help with the Harakari PID controller in Cleanflight.  Perhaps he could team up with someone who is familiar with the Cleanflight code to help with GPS code for Cleanflight.

http://www.multiwii.com/forum/viewtopic.php?f=23&t=3524&start=1400

https://github.com/Crashpilot1000/TestCode3

**Comment by jamieharwood** [comment]
I would vote for this. In my opinion the only failing of CleanFlight is it's GPS capabilities. I'm actually considering moving to the Pixhawk just for the GPS and Lidar functions but I would rather not because I really like the community around Cleanflight!!!

**Comment by dascons** [comment]
Why not both? Haha APM and CF, that's what I do

**Comment by MrMap** [comment]
I am a newcomer to the Cleanflight realm. Like it! 
Intend to use it with quad-racers but would also appreciate good GPS based features.
I hope new CF hardware platforms will have better support for multiple serial ports than the Naze32 I started out with... I am struggling to make it recognize the GPS at all...

Anyhow, my point, and I hope it is not tabu to say this in the CL open source community;
The ArduPilot / Arducopter developers has done a huge job to implement EKF (Extended Kalman Filter) and this is really a milestone in the navigation performance of related platforms.
http://en.wikipedia.org/wiki/Extended_Kalman_filter

Maybe you don´t need to re-invent the wheel completely?
Poking around in the Arducopter code repository might provide a jumpstart?

**Comment by dkisselev** [comment]
Arducopter's implementation is pretty complex, their have a very thorough navigation subsystem since the entire system is based around it. It's definitely a good reference, but from what I've seen, we're going for a much simpler final implementation, which might be easier to just develop from the core algorithm instead of trying to dumb down what they have.

They also wrote their to work in 3D if I remember correctly, which adds a lot of complexity to the kalman. I might be simpler for us to do it in 2D and work out another method for vertical navigation.

**Comment by digitalentity** [comment]
I agree with @dkisselev that Arducopter's navigation is complex. Arducopter's implementation is not unusable, but for CF it is much simpler and faster to re-invent the wheel and re-use some of existing code as well as some of @Crashpilot1000's code, which is indeed very good. 
Actually, new navigation code is mostly separated from position estimation. That is if somebody, sometime would implement a complete EKF with all bells and whistles, it would not require to re-write the code doing actual navigation.

**An update on my work with new navigation:**
I have mostly done writing code and now start to debug it. The new code compiles correctly but I've never bench-tested it. I'll need a couple of weeks to complete preliminary tests and hunt down some critical bugs. After that the code would be ready to be flight-tested.

The new code navigates in 3D space (lat, lon, altitude), but treats vertical (althold) and horizontal (poshold, rth) navigation separately. RTH is designed to climb to safe altitude, go home and land automatically. Poshold is split to POSHOLD_2D (holds onlt gps position) and POSHOLD_3D (actuially POSHOLD_2D + ALTHOLD). Poshold should now behave much like NAZA's "GPS Cruize" mode - pilot controls flight velocity, navigation code convert this to required attitude and throttle. If sticks are released - the craft stays where it is.

**Comment by digitalentity** [comment]
**Another update on my progress with rewriting navigation stack:**
Preliminary tests were good and I took a risk of flight-testing new NAV_ALTHOLD mode today. Weather conditions were not very good, wind was rather strong (about 7-10 mph) with strong gusts, also vertical flows were not stable. Did 4 or 5 short flights and recorded a small video. In my oppinion, new NAV_ALTHOLD code behaves much better than the legacy BARO mode. My quad was holding altitude perfectly, I was also able to control vertical speed (climb rate) with up to a few cm/s accuracy. I even did all landings without touching my throttle stick - descent was very smooth even near the ground.
Please see the video: https://youtu.be/7KJojPSE1PU

**Comment by marksev1** [comment]
Looks good since you were using such a noisy baro (bmp180), thats good news for my even older bmp085 hehe. Was this using some altitudehold code ported from harakiri, or without?

**Comment by digitalentity** [comment]
@marksev1, the althold part of nav-rewrite consist of my own code, legacy CF altitudehold code and CrashPilot's code, all mixed up, refactored and cleaned up. The new code also uses modified IMU (that's accelerometer + baro for climb rate)

**Comment by hydra** [comment]
@digitalentity Looking good.  What branch is your code in currently, i would like to see the changes.

**Comment by digitalentity** [comment]
That would be navigation-rewrite branch:
https://github.com/digitalentity/cleanflight/tree/navigation-rewrite?files=1

**Comment by digitalentity** [comment]
Did more testing today. POSHOLD does not work (digitalentity/cleanflight#3), ALTHOLD need better default PIDs (digitalentity/cleanflight/issues/4). I'll try to get ALTHOLD tuned better by myself, but I'll appreciate any help with this matter.

**Comment by digitalentity** [comment]
I've finally got new ALTHOLD and POSHOLD working. Please have look at this video: https://youtu.be/sGx-qlp0mO4

**Comment by dkisselev** [comment]
That's awesome! Does that mean that navigation also works if you set the position-to-hold to a far away spot?

**Comment by JohnOCFII** [comment]
Looks very good!  Next time, try to find a flag or something similar to have in the background to show the wind and how the craft handles it.

**Comment by digitalentity** [comment]
@dkisselev it should work, however there will be a special NAV_WP mode for that. The primary goal of PH is to get to target (hold) position as fast as possible without limiting velocity or acceleration, while NAV_WP would try to get smooth acceleration/deceleration.
@JohnOCFII that's a good idea. I'll try to get a flag and make another video of a flight test.

**Comment by digitalentity** [comment]
An update on my progress with navigation-rewrite.
ALTHOLD and POHOLD mostly work. Code is now being flight tested, new issues are being found and (hopefuly) fixed - see https://github.com/digitalentity/cleanflight/issues.
For the next week or two weather will be unflyable so I would appreciate any help with testing the code. Code review, ideas on improving the code (may be even PRs) are also very welcome!

**Comment by digitalentity** [comment]
@JohnOCFII, please see this video https://youtu.be/vMZLdUjpBnY - wind was strong (about 15 mph, gusts of up to 20 mph), code was imperfect and poorly tuned, but copter was handling the wind ok.

**Comment by JohnOCFII** [comment]
Very impressive (left a YouTube comment too)!  Once you think it is stable enough for testing, I could test it with my 260-class quad with a Naze32 clone (DragonFly32 Pro).

**Comment by Redshifft** [comment]
Sounds like might be testing all of this again soon :) , good to see you are trying to clean up some of those PID's that are not PID's :D

**Comment by digitalentity** [comment]
@JohnOCFII, I hope the weather this week will allow me to test new modifications to my code. Since that video I've made some major changes to PH code (ALTHOLD is affected as well). Hopefully I didn't break anything.
After I flight-test my latest changes the code will be ready for anybody to test in PH/ALTHOLD modes.

**Comment by vratinov** [comment]
Hi digitalentity,

I've been following your development and was wondering if you have anything new after your testing? I'd love to see good GPS hold for cleanflight.

thank you!

**Comment by csurf** [comment]
what's the status of RTH minimum height? I have a quad build on complete standby waiting for the CF GPS stuff to move ahead and include this feature. What's the hold-up and ETA on getting this done?

**Comment by FLDataTeK** [comment]
I've been testing CF on my 500mm Deadcat quad the past week and so far its doing pretty well, PH and RTH seem to function good with the exception of a tiny bit of toilet bowling but I think I need to re calibrate a bit to resolve that one. 

I am really interested in the minimum height on the RTH also because I live in Florida where everything is flat and we have 100 foot pine trees everywhere so I use RTH more for a oh crap switch more than anything. 

You can see at around 3:45 when I put it in PH and it finally has a 3D GPS lock it holds pretty good but you can see the toilet bowling on the RTH. 

https://youtu.be/ukY8r3ObCJg?t=2m25s

Here was a 1/2 mile flight I made today with the setup.. Flies pretty well overall. 

https://youtu.be/mU1bI4X14GI

Also I am able to test whatever you need..

**Comment by digitalentity** [comment]
@vratinov 
Thanks for your interest. There was alot of changes after latest testing. ALTHOLD and PH are mostly working. Currently I'm working on new default PIDs - see https://github.com/digitalentity/cleanflight/issues/4.
@csurf 
RTH is being developed and tested now. The hold-up is lack of time to flight-test the changes (and sometimes bad weather). So no ETA at the moment, might be tomorrow, might be couple of months.
@FLDataTeK
New navigation code is a complete rewrite, almost nothing of the current PH and RTH code remains. So it's not about adding a minimum height to RTH, it's about getting a whole new navigation stack to work. You are welcome to participate in development and testing. New default PIDs are a requirement before anything else can be moved forward. 

Those, who interested in testing, please grab the latest HEX (currently cleanflight_NAZE.navigation-rewrite.Aug12) here: https://www.dropbox.com/sh/4r5374mos7471q2/AACppn0Mphw45p-4TJ7lZLlOa?dl=0
**A BIG WARNING FOR TESTERS! The current code is in "beta" stage, it might not work as expected, so execute extreame caution and be prepared to switch to manual control at all times.
Also, a cood compass calibration is a "must do" thing. NAV_PH mode won't activate **

**Comment by FLDataTeK** [comment]
Thanks @digitalentity

**Comment by JohnOCFII** [comment]
@digitalentity Is the Hex above the release version of 1.9.0 with your changes, or is it the current master with your changes, or something in between?  I hope to test it this weekend.

**Comment by digitalentity** [comment]
@JohnOCFII 
My code is currently is based on 463437fb458fb6c81a29176a0f8c4b11ce8b3cc5 - that would be master as of Aug, 4. I'll rebase on current master when code is tested better or if master makes some serious changes. Please try and report your findings. Thanks in advance!

**Comment by digitalentity** [comment]
A new hex file is available - `cleanflight_NAZE.navigation-rewrite.Aug13.hex`
Now new PID and config defaults. Did several flights with this hex and new PIDs - all seem to work fine on my setup. Please test and report your findings.

**Comment by csurf** [comment]
@digitalentity, where can I take a look at your code changes? I don't personally feel comfortable blindly loading hexfiles from unknown branches/sources... I'd rather take a look at the code behind it first.

**Comment by dkisselev** [comment]
https://github.com/digitalentity/cleanflight/tree/navigation-rewrite

**Comment by digitalentity** [comment]
The NAV code itself resides in https://github.com/digitalentity/cleanflight/tree/navigation-rewrite, but hex-files are built from https://github.com/digitalentity/cleanflight/tree/experimental-build branch.
Experimental built includes ltm-telemetry-osb branch as I am using LTM-based OSD on my copter and also disable baro EOC interrupt on NAZE target as I fly a Naze Acro with a separate baro.

**Comment by csurf** [comment]
@digitalentity,
ok, I'm looking at your code, and it appears that you (or someone else) have already added the hold&climb routine during RTH init. 

I can see in the code:
                case NAV_RTH_STATE_CLIMB_TO_SAVE_ALTITUDE:
                    if (actualPosition.altitude > 1000) {
                        // Set target position to home and calculate original bearing
                        setNextWaypointAndCalculateBearing(homePosition.coordinates[LAT], homePosition.coordinates[LON], actualPosition.altitude);
                        navRthState = NAV_RTH_STATE_HEAD_HOME;
                    }

has this been tested?
The only thing I'd like to change is to make the RTH height a CLI-configurable value, instead of a fixed, hard-coded value (1000 meters in this case, or is it cm?). Thoughts?

I've setup a test quad and loaded your hex, hopefully I'll be able to test in the next few days.

**Comment by digitalentity** [comment]
@csurf 
I've done a few preliminary tests of this code - seem to work fine. Making RTH altitude a configurable value is work in progress - please see https://github.com/digitalentity/cleanflight/issues/10#issuecomment-130547841
BTW, I am now doing some refactoring and would appreciate a code review if you have some time to spare on this task.

**Comment by csurf** [comment]
@digitalentity,
ok, I took a look at the comment you linked to, and your plan to implement a multi-mode RTH height seems to make sense. The only difference in each case seems to be what altitude value will be set during the RTH_INIT / "climb&hold" phase.  

Perhaps it would make sense to simply have two CLI variables, something like RTH_ALT_MODE and RTH_ALT_VAL. RTH_ALT_MODE would simply be a flag to indicate which climb mode to implement, and RTH_ALT_VAL would be the actual height value to use (if necessary). 

RTH_ALT_MODE=
0: Keep current altitude, head straight home (RTH_ALT_VAL ignored)
1: Climb to RTH_ALT_VAL altitude above launch point, head home
2: Track the highest altitude during whole flight, climb there, head home (RTH_ALT_VAL ignored)
3: Climb to current altitude + RTH_ALT_VAL, head home

Upon initiating RTH, a switch-case would be executed in order to determine what altitude value to climb to based on the RTH_ALT_MODE/VAL values.

How does all of that sound? feedback is welcome.

as far as refactoring&review, sure, just let me know a few more specifics of what you'd like me to look at. FYI, not really a big deal, but I have seen a few spelling errors throughout the code in variable names. I could go through and fix those if necessary.

**Comment by digitalentity** [comment]
@csurf 
ref RTH - that's exactly what I am going to do.
ref code review - I'll let you know when I'm done with most of the current changes, rebasing and squashing lots of commits. Spelling errors are no surprise to me as I don't use English language on a daily basis :smile:

**Comment by csurf** [comment]
@digitalentity,
ok, sounds good.
What can I do to help speed up the process of getting the configurable options implemented? I could write up some code, but I'm not sure if you've already got something ready or are working on something. I'd just need to figure out how to add two new CLI config variables, and then do all of the monkey work of getting those patched in. Let me know how you'd like to proceed, as I really need to get this working ASAP.

**Comment by digitalentity** [comment]
@csurf 
Please see bda5a69f4077a914b0d007ba043a858bd7acfad6. Code is completely untested. Modes are defined as following:
```
    NAV_RTH_NO_ALT          = 0,            // Maintain current altitude
    NAX_RTH_EXTRA_ALT       = 1,            // Maintain current altitude + predefined safety margin
    NAV_RTH_CONST_ALT       = 2,            // Climb to predefined altitude
    NAV_RTH_MAX_ALT         = 3,            // Track maximum altitude and climb to it when RTH
```
Currently to get this fully flight ready I need lots of testing to close the following:
1) Find right values for PIDs: ALT, VEL, POS, POSR (defaults seem to work for me, but whey should be acceptable for other setups as well)
2) Find right value for `nav_gps_cf` CLI variable (probably somewhere around 0.2-0.6)
3) Play with `nav_pterm_cut_hz` and `nav_dterm_cut_hz` to see if we even need them
4) Make sure `nav_min_rth_distance` is working as it should
5) Take a copter without a barometer, but with GPS. Test that POSHOLD works,
6) Take a copter without GPS. Test that ALTHOLD works.
7) Test RTH. It should work even for setups without barometer (less altitude control).
Any assistance is greatly appreciated! Please fly and open new issues here: https://github.com/digitalentity/cleanflight/issues

**Comment by sppnk** [comment]
@digitalentity
 Can  NAV_RTH_CONST_ALT   mode "climb" to a predefined  lower altitude? May be interesting for future fixed wing implementation.

**Comment by digitalentity** [comment]
@sppnk 
Please see 1d145c806c539afb09dc4e2cd73442b22ff6b7ae. Introduced new NAV_RTH_AT_LEAST_ALT mode (now a config default) which will only climb to at least defined altitude. All other modes will descend as well (if needed).

**Comment by sppnk** [comment]
ok, thx

**Comment by digitalentity** [comment]
New build of nav-rewrite is ready for testing: https://github.com/digitalentity/cleanflight/issues/1#issuecomment-135613961

**Comment by digitalentity** [comment]
I'm happy to announce a very first public beta of my navigation-rewrite for Cleanflight featuring better altitude and position hold, better RTH and GPS-assisted flight modes. Please have a look at my post at [RCG](http://www.rcgroups.com/forums/showthread.php?t=2495732)

**Comment by madixmadax** [comment]
Looks like the new navigation was included in master.

**Comment by gregswf** [comment]
Hi,
Used your firmware yesterday.
I think it is excellent, thank you for doing it. 
The altitude hold works very well with default PID settings.
The position hold overshoots so I think I have to change the PID settings, I also have to put my mag on a post and tweak it a bit to make sure it is in line with the frame.
My build:
F450 frame, Naze32 full with external HMC5883L mag, neo-6m GPS, Emax 2213-935kv motors, hobbywing Xrotor ESC's, Graupner eprop 9x5 props, Turnigy nanotech 5000mAh battery, FRsky radio.

**Comment by Ravid824** [comment]
Is any of the new rewrite code is part of the latest stable cleanflight 1.10 released on OCT?
What is the best stable firmware to be used from 
 https://www.dropbox.com/sh/4r5374mos7471q2/AACppn0Mphw45p-4TJ7lZLlOa?dl=0

**Comment by digitalentity** [comment]
@Ravid824 
No, none of that code is in the Cleanflight yet. There is no "stable" firmware in my dropbox - all betas, all with different bugs. Feel free to join the testing and discussion here http://www.rcgroups.com/forums/showthread.php?t=2495732

**Comment by phijoule** [comment]
Hi @digitalentity . I am totally new to github and Naze32. I have setup a few 250 scale quad of my friends running on Naze 32 with GPS and OSD. Using the latest Cleanflight 1.9.0, I was able to see the GPS count (8 to 9 sats) but were not able to use the PH (Altitude Hold isn't working too). 

Could you do a quick demo on youtube on import your hex file into the Naze32 board. I am assuming it would be similar to uploading stable Naze32 hex file into the board, Yes? Does the hex file contains all the default value for GPS PH and Altitude Hold? I presume it would work with Flip32+ board too right?

Glad to see that you are working on the RTL, but for now PH and Alt.Hold is good enough for me to start with. Thanks

**Comment by Ravid824** [comment]
hydra, do you have any plans to include Improve GPS Position Hold (if any) in the coming releases?
This issue is one of the biggest CF limitation which is used for non racers users (such photography and more).

**Comment by JohnOCFII** [comment]
@Ravid824 Take a look at the improved navigation work being done in the iNav version of CleanFlight.  I believe there is still a desire to roll those improvements back in to CleanFlight, but this will show you the current state:  http://www.rcgroups.com/forums/showthread.php?t=2495732

**Comment by Ravid824** [comment]
I am aware of iNAV development efforts and I wish that Hydra has any commitment to include it soon in the official CF builds.

**Comment by TinoGroteloh** [comment]
#1621
